File: algorithms/javascriptalgo/leetcode/number501-600/number541-550/number543/number424/diameter-of-binary-tree.test.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
const diameterOfBinaryTree = function(root) {
  let distance = 0;

  const dfs = (node) => {
    if (!node) return 0;

    let left = dfs(node.left);
    let right = dfs(node.right);

    distance = Math.max(distance, left + right);

    return Math.max(left, right) + 1;
  }

  dfs(root);
  return distance;
};

File: algorithms/javaalgo/src/test/java/leetcode/number501to600/number541to550/number543/DiameterOfBinaryTreeTest.java
package leetcode.number501to600.number541to550.number543;

import leetcode.util.TreeNode;

public class DiameterOfBinaryTreeTest {

    int distance = 0;

    public int diameterOfBianryTree(TreeNode node) {
        dfs(node);
        return distance;
    }

    private int dfs(TreeNode node) {
        if (node == null) {
            return 0;
        }

        int left = dfs(node.left);
        int right = dfs(node.right);

        distance = Math.max(distance, left + right);

        return Math.max(left, right) + 1;
    }
}
